### PR TITLE
Upgrade libtorch to 1.8

### PR DESCRIPTION
**Pull request by junjihashimoto**
This is a PR to upgrade libtorch to pytorch-1.8.
https://pytorch.org/blog/pytorch-1.8-released/

- [x] Upload libtorch binaries to [libtorch-binary-for-ci](https://github.com/hasktorch/libtorch-binary-for-ci/releases/tag/1.8.0).
- [x] Update libtorch's links in pytorch-world (https://github.com/stites/pytorch-world/pull/26).
- [x] Update deps/get-deps.sh
- [x] Run codegen and update source code of libtorch-ffi package.
- [x] Update Torch.Functional.Internal of hasktorch package.  (I modify some codes by hand so that it can be compiled.
)
- [x] Test

The behavior of both median-function and embedding-function has changed.

https://github.com/pytorch/pytorch/issues/53368 shows changes to the behavior of embedding-function.

```
/home/junjihashimoto/git/hasktorch/hasktorch/src/Torch/Typed/Functional.hs:419: failure in expression `median @0 @KeepDim t'
expected: (Tensor Float [1,2] [[ 3.0000   ,  1.0000   ]],Tensor Int64 [1,2] [[ 1,  0]])
 but got: (Tensor Float [1,2] [[ 3.0000   ,  1.0000   ]],Tensor Int64 [1,2] [[ 1,  2]])
                                                                                   ^

/home/junjihashimoto/git/hasktorch/hasktorch/src/Torch/Typed/Functional.hs:2811: failure in expression `dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[[Float]]]) $ t'
expected: (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[2.0,2.0]]]))
 but got: (Float,([4,1,2],[[[0.0,0.0]],[[3.0,3.0]],[[0.0,0.0]],[[2.0,2.0]]]))
```

### CODE REVIEW

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 1]
```diff
@@ -2805,16 +2808,27 @@ type family PaddingIdxCheck (idx :: Maybe Nat) (numEmbeds :: Nat) :: Constraint
...
 -- t :: Tensor '( 'D.CPU, 0) 'D.Float '[4, 1, 2]
+-- -- libtorch 1.7
+-- -- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[[Float]]]) $ t
+-- -- (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[2.0,2.0]]]))
+-- -- 
+-- -- libtorch 1.8
+-- -- The behavior of libtorch 1.8 changes. See https://github.com/pytorch/pytorch/issues/53368
 -- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[[Float]]]) $ t
--- (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[2.0,2.0]]]))
+-- (Float,([4,1,2],[[[0.0,0.0]],[[3.0,3.0]],[[0.0,0.0]],[[2.0,2.0]]]))
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Functional.hs:2821]: For now, modify test codes to pass the change of behavior.

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 2]
```diff
@@ -2805,11 +2809,18 @@ type family PaddingIdxCheck (idx :: Maybe Nat) (numEmbeds :: Nat) :: Constraint
...
+-- 
+-- -- libtorch 1.7
+-- -- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[[Float]]]) $ t
+-- -- (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[2.0,2.0]]]))
+-- -- 
+-- -- libtorch 1.8
+-- -- The behavior of libtorch 1.8 changes. See https://github.com/pytorch/pytorch/issues/53368
 -- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[[Float]]]) $ t
--- (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[2.0,2.0]]]))
+-- (Float,([4,1,2],[[[1.0,1.0]],[[3.0,3.0]],[[1.0,1.0]],[[0.0,0.0]]]))
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Functional.hs:2823]: The change of behavior.
https://github.com/pytorch/pytorch/issues/53368

**Review by tscholak — APPROVED** [review]
fantastic work @junjihashimoto!

Have you tested rocm?

**Comment by junjihashimoto** [comment]
> fantastic work @junjihashimoto!
> 
> Have you tested rocm?

**Comment by junjihashimoto** [comment]
@tscholak 
I've not tested it yet.
But I'll do it on aws g4ad instance.

**Comment by junjihashimoto** [comment]
rocm's binary needs  `ghc-options:       -optc-D_GLIBCXX_USE_CXX11_ABI=0`. It is Pre-cxx11 ABI.

image

**Comment by junjihashimoto** [comment]
I've fixed the linkage program of libtorch-rocm, but libnuma throws following error.
image

I should cut out rocm into another PR.

**Comment by junjihashimoto** [comment]
When we don't use nix, this error does not occur.
image

**Comment by junjihashimoto** [comment]
This PR does not support rocm for nix.
It is possible to build some examples without nix.

**Comment by junjihashimoto** [comment]
There are no remaining issues, so merge this pr.

**Pull request merged**